internal/controller/standalone_pgadmin/watches.go

```diff
@@ -0,0 +1,115 @@
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package standalone_pgadmin
+
+import (
+	"context"
+
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+
+	"github.com/crunchydata/postgres-operator/pkg/apis/postgres-operator.crunchydata.com/v1beta1"
+)
+
+// watchPostgresClusters returns a [handler.EventHandler] for PostgresClusters.
+func (r *PGAdminReconciler) watchPostgresClusters() handler.Funcs {
+	handle := func(cluster client.Object, q workqueue.RateLimitingInterface) {
+		ctx := context.Background()
+		for _, pgadmin := range r.findPGAdminsForPostgresCluster(ctx, cluster) {
+
+			q.Add(ctrl.Request{
+				NamespacedName: client.ObjectKeyFromObject(pgadmin),
+			})
+		}
+	}
+
+	return handler.Funcs{
+		CreateFunc: func(e event.CreateEvent, q workqueue.RateLimitingInterface) {
+			handle(e.Object, q)
+		},
+		UpdateFunc: func(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+			handle(e.ObjectNew, q)
+		},
+		DeleteFunc: func(e event.DeleteEvent, q workqueue.RateLimitingInterface) {
+			handle(e.Object, q)
+		},
+	}
+}
+
+// watchForRelatedSecret handles create/update/delete events for secrets,
+// passing the Secret ObjectKey to findPGAdminsForSecret
+func (r *PGAdminReconciler) watchForRelatedSecret() handler.EventHandler {
+	handle := func(secret client.Object, q workqueue.RateLimitingInterface) {
+		ctx := context.Background()
+		key := client.ObjectKeyFromObject(secret)
+
+		for _, pgadmin := range r.findPGAdminsForSecret(ctx, key) {
+			q.Add(ctrl.Request{
+				NamespacedName: client.ObjectKeyFromObject(pgadmin),
+			})
+		}
+	}
+
+	return handler.Funcs{
+		CreateFunc: func(e event.CreateEvent, q workqueue.RateLimitingInterface) {
+			handle(e.Object, q)
+		},
+		UpdateFunc: func(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+			handle(e.ObjectNew, q)
+		},
+		// If the secret is deleted, we want to reconcile
+		// in order to emit an event/status about this problem.
+		// We will also emit a matching event/status about this problem
+		// when we reconcile the cluster and can't find the secret.
+		// That way, users will get two alerts: one when the secret is deleted
+		// and another when the cluster is being reconciled.
+		DeleteFunc: func(e event.DeleteEvent, q workqueue.RateLimitingInterface) {
+			handle(e.Object, q)
+		},
+	}
+}
+
+//+kubebuilder:rbac:groups="postgres-operator.crunchydata.com",resources="pgadmins",verbs={list}
+
+// findPGAdminsForSecret returns PGAdmins that have a user or users that have their password
```

Comment:
:question: Is this comment correct? If I'm understanding correctly, `matching` returns the PGAdmins that have a user with a password stored in a Secret, but wouldn't return for users without a `PasswordRef`. Maybe I'm missing something?

Reply:
I think I see what I was misunderstanding. I read it as "have a user" or "have users that have their password stored in the Secret", but the comment means "returns PGAdmins that have one or more users that have their password stored in the Secret". If that's correct, I'm good with the comment as is.